ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java

```diff
@@ -293,36 +293,23 @@ private Callback<Long> fetchStitchRequestBodyCallback(RetainingAsyncWritableChan
      * After {@link Router#putBlob} finishes, call {@link IdConverter#convert} to convert the returned ID into a format
      * that will be returned in the "Location" header.
      * @param blobInfo                       the {@link BlobInfo} to use for security checks.
-     * @param propertiesPassedInRouterUpload the {@link BlobProperties} instance that is passed to Router during upload
      * @return a {@link Callback} to be used with {@link Router#putBlob}.
      */
-    private Callback<String> routerStitchBlobCallback(BlobInfo blobInfo,
-        BlobProperties propertiesPassedInRouterUpload) {
-      return buildCallback(frontendMetrics.putRouterStitchBlobMetrics, blobId -> {
+    private Callback<String> routerStitchBlobCallback(BlobInfo blobInfo, BlobProperties propertiesForRouterUpload) {
+      return buildCallback(frontendMetrics.putRouterStitchBlobMetrics, convertedBlobId -> {
         // The actual blob size is now present in the instance of BlobProperties passed to the router.stitchBlob().
         // Update it in the BlobInfo so that IdConverter can add it to the named blob DB
-        blobInfo.getBlobProperties().setBlobSize(propertiesPassedInRouterUpload.getBlobSize());
-        idConverter.convert(restRequest, blobId, blobInfo.getBlobProperties(), idConverterCallback(blobInfo, blobId));
-      }, uri, LOGGER, deleteDatasetCallback);
-    }
-
-    /**
-     * After {@link IdConverter#convert} finishes, call {@link SecurityService#postProcessRequest} to perform
-     * request time security checks that rely on the request being fully parsed and any additional arguments set.
-     * @param blobInfo the {@link BlobInfo} to use for security checks.
-     * @return a {@link Callback} to be used with {@link IdConverter#convert}.
-     */
-    private Callback<String> idConverterCallback(BlobInfo blobInfo, String blobId) {
-      return buildCallback(frontendMetrics.putIdConversionMetrics, convertedBlobId -> {
+        blobInfo.getBlobProperties().setBlobSize(propertiesForRouterUpload.getBlobSize());
```

Comment:
This is too late, the blob properties().setBlobSize should be called before idconverter, idconverter has to get the blob size and write it to named blob database, that's why we need the blob size. we have to do this before the idconverter.